### PR TITLE
SFA-414 - Java Implementation of the SDK Manager 

### PR DESCRIPTION
**Pull request by BSaunders95**
Java implementation of the SDK Manager, used to return an instance of the `api-sdk-java` which follows an OAuth authentication protocol if a user is logged in; the library fulfils 2 purposes:
- Access the session to determine whether a user is logged in or not. Use that logic to return an SDK instance with a suitable authentication protocol.
- Provide a `SessionTokenRefreshHandler` which is executed when making API requests with an expired OAuth token, to update the access token on the session.

This is part of the splitting of the session handler. Resolves [SFA-414](https://companieshouse.atlassian.net/browse/SFA-414)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,43 @@
...
+If you want to just build and install the library then use the dev profile which skips unit tests:
+```
+mvn clean install -Pdev
+```
+
+The library can be imported as a maven dependency:
+```
+<dependency>
+    <groupId>uk.gov.companieshouse</groupId>
+    <artifactId>sdk-manager-java</artifactId>
```
Comment (jsimons1989) [inline README.md]: Can you include a simple code example of how to use the library please?

File: README.md [hunk 2]
```diff
@@ -1 +1,43 @@
...
+
+## Getting started
+The library is built using maven.  For example, to build, install and run unit tests in the local maven repo, run the following:
+```
+mvn clean install
+```
+
+If you want to just build and install the library then use the dev profile which skips unit tests:
+```
+mvn clean install -Pdev
```
Comment (jsimons1989) [inline README.md]: I would advise against including instructions to bypass the unit tests in the README.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                tokenResponse.setAccessToken(sessionAccessToken.getToken());
+                tokenResponse.setRefreshToken(sessionAccessToken.getRefreshToken());
+            }
+
+            TokenRefreshHandler handler = new SessionTokenRefreshHandler(session);
+
+            httpClient = new OAuthHttpClient(tokenResponse, getClientId(), getClientSecret(),
+                    handler);
+        } else {
+            httpClient = new ApiKeyHttpClient(CHS_API_KEY);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java]: Shouldn't the parameter be the API key in the env properties?

File: README.md (hunk 2 above)
Comment (cburns93) [reply README.md]: It's just to stop devs having to comment out the JUnits that are failing just so they can get their code working and compiling. I face problems a lot of the time where I want to compile code to test functionality but cannot do an install due to the tests failing. This profile is just for building purposes. Realistically, in the Make and Jenkins problem, the unit tests will be run.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java (hunk 3 above)
Comment (cburns93) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java]: Do you mean creating another file called `env.properties`?

Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java]: We should be pulling the `CHS_API_KEY` from `global_env`, sorry I should have probably been clearer

Comment (jsimons1989) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java]: Sorry if I was unclear.

You have the key defined as the name of the env variable:
```
private static final String CHS_API_KEY = "CHS_API_KEY";
```

then you are passing the key name into the HTTP client. Shouldn't you be resolving the env variable?
```
httpClient = new ApiKeyHttpClient(CHS_API_KEY);
```

File: README.md (hunk 2 above)
Comment (jsimons1989) [reply README.md]: If the unit tests are failing, then the functionality wouldn't be working either. Unit tests should be passing first before attempting to build and test the integration of the units. We should be trying to follow the TDD approach.

File: README.md (hunk 1 above)
Comment (cburns93) [reply README.md]: Yep, done. Let me know if it is enough

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java (hunk 3 above)
Comment (cburns93) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManagerImpl.java]: I understand now, fix has been committed

File: README.md (hunk 2 above)
Comment (cburns93) [reply README.md]: That - in theory - is a good idea. But realistically, when it comes down to it when the unit tests are written, they are written last. When people develop new functionality, they test to see if it works first and the first step of that is building the project. If they where to fix the countless JUnit tests that where failing, they could spend a lot of time on that alone - eventually to find out that the code they written didn't work anyways and all the time spent fixing the junit tests have been wasted. Adding the `dev` profile is just a way of saving developer time in disabling all Junit tests. Which is what they do anyways. I've added a comment in the README.md that advises people to run the test profile before they commit.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    private static final String X_REQUEST_ID_HEADER = "x-request-id";
+    private static final String CHS_API_KEY = "CHS_API_KEY";
+    private static final String OAUTH2_CLIENT_ID = "OAUTH2_CLIENT_ID";
+    private static final String OAUTH2_CLIENT_SECRET = "OAUTH2_CLIENT_SECRET";
+
+    private static EnvironmentReader environmentReader = new EnvironmentReaderImpl();
+
+    private ApiClientManager() {
+        // private constructor set up to prevent instantiation
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java]: This isn't needed.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    private static String getClientId() {
+        return environmentReader.getMandatoryString(OAUTH2_CLIENT_ID);
+    }
+
+    private static String getClientSecret() {
+        return environmentReader.getMandatoryString(OAUTH2_CLIENT_SECRET);
+    }
+
+    private static String getCHSApiKey() {
+        return environmentReader.getMandatoryString(CHS_API_KEY);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java]: What happens if a mandatory string is not set? Will an error be logged making it clear which environment variable is missing?

File: src/main/java/uk/gov/companieshouse/sdk/util/SessionTokenRefreshHandler.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+public class SessionTokenRefreshHandler implements TokenRefreshHandler {
+
+    private Session session;
+
+    public SessionTokenRefreshHandler(Session session) {
+        this.session = session;
+    }
+
+    /**
+     * handleRefresh is fired when making an API call against an expired OAuth token.
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/sdk/util/SessionTokenRefreshHandler.java]: The JavaDoc shouldn't be defining the use case for a method (handleRefresh is fired when making an API call). It should be enabling the developer to decide whether the method is providing the functionality they need.

File: src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    public void setup() {
+        session = new SessionImpl(store, "", new HashMap<>());
+
+        ServletRequestAttributes attrs = new ServletRequestAttributes(request);
+        RequestContextHolder.setRequestAttributes(attrs);
+    }
+
+    @Test
+    @DisplayName("Test a successful OAuth client is returned")
+    public void get_oauth_api_client() {
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java]: Method naming is not following Java conventions. See the examples in https://junit.org/junit5/docs/current/user-guide/

File: src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        createSessionData();
+        ApiClient apiClient = ApiClientManager.getSDK();
+
+        // tests that the api client returned by getSDK() is the OAuthHttpClient
+        assertEquals(OAuthHttpClient.class, apiClient.getHttpClient().getClass());
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java]: Comment is not needed. It is not adding any additional information to the assertion.

File: src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    @DisplayName("Confirm RequestId is generated if blank")
+    public void confirm_requestId_is_generated_if_blank() {
+        createSessionData();
+        when(request.getHeader(ApiClientManager.getXRequestIdHeader())).thenReturn("");
+
+        ApiClient apiClient = ApiClientManager.getSDK();
+        OAuthHttpClient oAuthHttpClient = (OAuthHttpClient) apiClient.getHttpClient();
+        String requestId = oAuthHttpClient.getRequestId();
+
+        assertEquals(20, requestId.length());
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/sdk/ApiClientManagerTest.java]: Please place the duplicated number in a constant to make it easier to change.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java (hunk 5 above)
Comment (cburns93) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java]: Yes, this is a test I but somehow it seems I removed it. I will add back in. But yes the error would be something along the lines of this during the build:
```[ERROR] get_api_key_client  Time elapsed: 0.284 s  <<< ERROR!
uk.gov.companieshouse.environment.exception.EnvironmentVariableException: Environment variable missing: CHS_API_KEY
	at uk.gov.companieshouse.sdk.ApiClientManagerTest.get_api_key_client(ApiClientManagerTest.java:85)```

And at runtime the stacktrace would report:
```java
uk.gov.companieshouse.environment.exception.EnvironmentVariableException: Environment variable missing: CHS_API_KEY

	at uk.gov.companieshouse.environment.impl.EnvironmentReaderImpl.verifyNotNull(EnvironmentReaderImpl.java:97)
	at uk.gov.companieshouse.environment.impl.EnvironmentReaderImpl.getMandatoryString(EnvironmentReaderImpl.java:30)
	at uk.gov.companieshouse.sdk.manager.ApiClientManager.getCHSApiKey(ApiClientManager.java:99)
	at uk.gov.companieshouse.sdk.manager.ApiClientManager.getSDK(ApiClientManager.java:62)
	at uk.gov.companieshouse.sdk.ApiClientManagerTest.get_api_key_client(ApiClientManagerTest.java:85)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:513)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:113)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:121)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:121)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:121)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:121)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

File: src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        // Construct pre-existing session with an old access token and expiration period
+        final String oldAccessToken = "Foo";
+        final String oldExpirationPeriod = "123456789";
+
+        Map<String, Object> accessTokenInput = new HashMap<>();
+        accessTokenInput.put(AuthFilter.SESSION_KEY_ACCESS_TOKEN, oldAccessToken);
+        accessTokenInput.put(AuthFilter.SESSION_KEY_EXPIRES_IN, oldExpirationPeriod);
+
+        // Embed the access token into a sign-in data map
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java]: These comments are not needed.

File: src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java (hunk 4 above)
Comment (cburns93) [reply src/main/java/uk/gov/companieshouse/sdk/manager/ApiClientManager.java]: How come?

File: src/main/java/uk/gov/companieshouse/sdk/util/SessionTokenRefreshHandler.java (hunk 6 above)
Comment (cburns93) [reply src/main/java/uk/gov/companieshouse/sdk/util/SessionTokenRefreshHandler.java]: I think @BSaunders95 was stuck in Golang comment mode :P

File: README.md (hunk 1 above)
Comment (jsimons1989) [reply README.md]: Yep :) a simple one was enough. Just didn't want developers having to read the code to find the command to get the SDK instance.

File: src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        sessionTokenRefreshHandler.handleRefresh(tokenResponse);
+
+        // Access the nested access token map from within session data
+        Map<String, Object> sessionDataOutput = session.getData();
+        Map<String, Object> signInDataOutput = (Map<String, Object>) sessionDataOutput.get(AuthFilter.SESSION_KEY_SIGN_IN_INFO);
+        Map<String, Object> accessTokenOutput = (Map<String, Object>) signInDataOutput.get(AuthFilter.SESSION_KEY_ACCESS_TOKEN);
+
+        // Assert access token has been updated
+        String refreshedAccessToken = (String) accessTokenOutput.get(AuthFilter.SESSION_KEY_ACCESS_TOKEN);
+        assertEquals(refreshedAccessToken, newAccessToken);
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java]: Are these the assertions the wrong way round? Aren't you confirming that the session has been updated with the new access token and expiry?

File: src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        Map<String, Object> signInDataOutput = (Map<String, Object>) sessionDataOutput.get(AuthFilter.SESSION_KEY_SIGN_IN_INFO);
+        Map<String, Object> accessTokenOutput = (Map<String, Object>) signInDataOutput.get(AuthFilter.SESSION_KEY_ACCESS_TOKEN);
+
+        // Assert access token has been updated
+        String refreshedAccessToken = (String) accessTokenOutput.get(AuthFilter.SESSION_KEY_ACCESS_TOKEN);
+        assertEquals(newAccessToken, refreshedAccessToken);
+
+        // Assert expiration period has been updated
+        Integer refreshedExpirationPeriod = (Integer) accessTokenOutput.get(AuthFilter.SESSION_KEY_EXPIRES_IN);
+        assertEquals(refreshedExpirationPeriod, Integer.valueOf(newExpirationPeriod));
```
Comment (jsimons1989) [inline src/test/java/uk/gov/companieshouse/util/SessionTokenRefreshHandlerTest.java]: These parameters also need to be switched around:
```java
assertEquals(Integer.valueOf(newExpirationPeriod), refreshedExpirationPeriod);
```

**Pull request merged**